Fixes #22847 - Use pf-react donut chart in facts page

lizagilman opened the pull request.

lizagilman committed:
Fixes #22847 - Use pf-react donut chart in facts page

ohadlevy commented:
@lizagilman thanks for the pr! @danseethaler can you please review?

danseethaler commented:
hey @lizagilman - you should be able to run `npm run lint` locally to fix up these HoundCI issues before you push. I've forgotten that lots of times 😀

I think it would be helpful to add a storybook to demo the chart changes you're making here. Let me know your thoughts on that and if you need any direction. I don't see any way to demo the changes currently.

Also, I don't see the `c3` dependency actually being used in the code. Is it coming from other components that you're pulling in? I'm new to the charting changes we're making so thanks for giving some background.

lizagilman commented:
Hi Dan,
I have two components here - the modal with the chart and the button that opens it
The modal body contains the c3 chart (implemented in `<PieChart> `component)
This replaces the flot/ jquey implementation of fact distribution charts
I added the modal component to storybook
Thanks!

danseethaler reviewed (approved):
This looks great. I tried out the storybook which works well. I am looking forward to see how these changes will be integrated into some pages. Looks great for now!

ohadlevy commented on webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js:
```diff
@@ -0,0 +1,55 @@
+      <Modal show={factChart.showModal} onHide={this.closeModal}>
+        <Modal.Header>
+          <Modal.Title>
+            <b>{__(`Fact distribution chart - ${factChart.title}`)}</b>
```
I think this needs to be extracted differently, probably using sprintf, see javascript section under http://projects.theforeman.org/projects/foreman/wiki/Translating

ohadlevy commented on webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js:
```diff
@@ -0,0 +1,55 @@
+}
+
+const mapStateToProps = (state, ownProps) => ({
+  factChart: ownProps.factChart || state.factChart,
```
when do you get ownProps with factChart vs state?

lizagilman replied:
this makes the componenet also resuable by just passing props without dealing with state
initially this was sort of a workaround to get storybook to display the component as we couldn't reach redux state in storybook
@sharvit

ohadlevy commented on webpack/assets/javascripts/react_app/components/factChart/index.js:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  render() {
+    return (<Button onClick={this.onClick}>View Chart</Button>);
```
please extract the string

waldenraines replied:
This is not translated.

ohadlevy commented on webpack/assets/javascripts/react_app/redux/actions/factCharts/index.js:
```diff
@@ -0,0 +1,29 @@
+  FACT_CHART_DATA_REQUEST,
+  FACT_CHART_DATA_SUCCESS,
+  FACT_CHART_DATA_FAILURE,
+  OPEN_MODAL,
```
Action names are global - OPEN/CLOSE are too generic names, it should be specific to the page / action

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/factChart/index.js:
```diff
@@ -0,0 +1,31 @@
+
+export default (state = initialState, action) => {
+  switch (action.type) {
+    case FACT_CHART_DATA_SUCCESS:
```
where are the implementation for FACT_CHART_DATA_FAILURE and FACT_CHART_DATA_REQUEST?

lizagilman replied:
pie chart component displays spinner while waiting and an error message when there's no data.
what should we do in request and failure here?

ohadlevy replied:
IMHO once the action are fired, than we should change the state, for example, on failure - set the correct title etc.
on loading, it should set the loaderStatus correctly etc.

ohadlevy reviewed (changes requested):
thanks @lizagilman i did an initial code review - looking good :-)

added some inline comments, and of course, looking forward to seeing tests :)

waldenraines commented on webpack/assets/javascripts/react_app/components/componentRegistry.js:
```diff
@@ -55,15 +56,17 @@ const componentRegistry = {
 };
 
-const coreComponets = [
+const coreComponents = [
```
Thanks for fixing that typo! 👍

waldenraines commented on webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js:
```diff
@@ -0,0 +1,55 @@
+    const error = (
+      <MessageBox
+        msg={__('No data available')}
+        icontype="error-circle-o"
```
I think we should use icons from [patternfly's icon list](https://www.patternfly.org/styles/icons/).  Perhaps `pficon-error-circle-o`?

lizagilman replied:
I thinks it's already a pf icon at `<MessageBox>`

waldenraines replied:
Oh I see.  I'm not a fan of how the `<MessageBox>` [restricts you](https://github.com/theforeman/foreman/blob/a5a6b5df465560729c86fa4799c1e9ad13941444/webpack/assets/javascripts/react_app/components/common/MessageBox.js#L11) to using `pficon-${iconType}` (as patternfly also recommends font awesome icons in some cases) but that certainly has nothing to do with this PR.

waldenraines commented on webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js:
```diff
@@ -0,0 +1,55 @@
+          <Loader status={factChart.loaderStatus}>{[chart, error]}</Loader>
+        </Modal.Body>
+        <Modal.Footer>
+          <Button onClick={this.closeModal}>Close</Button>
```
This is not translated.

ohadlevy commented on app/views/fact_values/index.html.erb:
```diff
@@ -27,4 +27,7 @@ end
 </table>
 
+<span id="fact-chart-modal"></span>
+<%= mount_react_component('FactChartModal', "#fact-chart-modal") %>
```
q: is this really required? e.g. if on Click you would open the modal, does it need a second component mounted already or can you simply open the modal and render your component inside?

ohadlevy replied:
bump?

waldenraines commented:
@lizagilman I wonder if this would be helpful https://github.com/patternfly/patternfly-react/pull/103

lizagilman commented:
updated to use single component (button+modal instead of one for button and one for modal)
added some more stories
some changes in reducer logic 
WIP tests

lizagilman commented:
add component and reducer tests
WIP action tests

lizagilman commented:
[test foreman]
[test katello]

lizagilman commented:
[test foreman]

lizagilman commented:
[test foreman]

lizagilman commented:
[test katello]

lizagilman commented:
@danseethaler / @ohadlevy can you take another look?

mmoll commented:
[test katello]
[test upgrade]

waldenraines commented:
[test katello]

amirfefer commented on .storybook/preview-head.html:
```diff
@@ -11,3 +11,12 @@
+<!-- workaround to get Jed into storybook -->
+<script>
+  window.Jed = {
+    sprintf: function(str, elem) {
```
nitpick: use es6 arrow function

amirfefer commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,85 @@
+}
+
+const mapStateToProps = (state, ownProps) => ({
+  factChart: ownProps.factChart || state.factChart,
```
You can map the specific dispalyModal state to props (in `mapsStateToProps` function), something like this:
```javascript
displayModal: state.factChart.displayModal[ownProps.id] || false,
```
after this change,  line 32 becomes redundant

amirfefer commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,85 @@
+  render() {
+    const { factChart, data: { id, title } } = this.props;
+
+    const showModal = factChart.displayModal[id];
```
its the same name as the action function, shouldn't it has a unique name?
or maybe map `displayModal` to a prop?

amirfefer commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,85 @@
+    const chartConfig = showModal ? getLargePieChartConfig({ data: factChart.chartData }) : false;
+
+    const chart = showModal ? (<Chart
+        className="c3-statistics-pie"
```
Why not using `<PieChart>` component?

lizagilman replied:
`<PieChart>` is `small` by default

amirfefer replied:
why not pass it as a prop?

amirfefer commented on webpack/assets/javascripts/services/ChartService.consts.js:
```diff
@@ -20,7 +20,7 @@ export const donutChartConfig = {
     show: true,
   },
-  legend: { show: false },
+  legend: { show: true },
```
shouldn't we add legend on large charts only ?

lizagilman replied:
looks ok to me. how does it look to you?

amirfefer replied:
Its impact small size charts as well

In dashboard  you have legend information twice:
![donut_chart](https://user-images.githubusercontent.com/11807069/34765170-79907976-f5f9-11e7-91f6-dc4f630165b8.png)

IMHO adding legend into a small size chart is a bit redundant, because its too small. instead we could click to enlarge it, and see it.

![statistics](https://user-images.githubusercontent.com/11807069/34765085-3fe51790-f5f9-11e7-8d0d-4ba4fe874f97.png)

lizagilman commented:
@amirfefer addressed all your comments

amirfefer commented on .storybook/preview-head.html:
```diff
@@ -11,3 +11,12 @@
+  const sprintfStr = (str, elem) => { return str } ;
+
+  window.Jed = {
+    sprintf:   sprintfStr
```
Does it work well when running stroybook? shouldn't it be a function?

lizagilman replied:
Works well, it is sort of a workaround to get Jed work in storybook

amirfefer commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,80 @@
+            </Modal.Footer>
+          </Modal>
+          :
+          false
```
Instead of if statement, you could use 
```javascript
modalToDisplay && <Component>
```

amirfefer commented on webpack/assets/javascripts/react_app/redux/reducers/index.js:
```diff
@@ -5,6 +5,8 @@ import statistics from './statistics';
 import notifications from './notifications/';
 import toasts from './toasts';
+import factChart from './factCharts/';
+
```
redundant line

lizagilman commented:
@amirfefer updated

lizagilman commented:
[test katello]

lizagilman commented:
@danseethaler any comments?

lizagilman commented:
@amirfefer ping

amirfefer commented on webpack/assets/javascripts/react_app/components/factCharts/factChart.fixtures.js:
```diff
@@ -0,0 +1,35 @@
+});
+
+export const modalOpenState = Immutable({
+  factChart: {
```
Similar repetitive blocks, could you dry it a bit?

amirfefer commented on .storybook/preview-head.html:
```diff
@@ -11,3 +11,12 @@
+  const sprintfStr = (str, elem) => { return str } ;
+
+  window.Jed = {
+    sprintf:   sprintfStr
```
nitpick: redundant space?

amirfefer commented on app/controllers/facts_controller.rb:
```diff
@@ -13,6 +13,7 @@ def show
     rescue
       data = @fact
     end
+    data[:values] = data[:values].map { |x| x.values }
```
what about something like this: `data[:values].map!(&:values)`?
`map!` replaces the element

amirfefer commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,79 @@
+    helpers.bindMethods(this, ['onClick', 'closeModal']);
+  }
+
+  onClick() {
```
could you pick a more informative name?

amirfefer commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,79 @@
+        smallChart = {false}
+      />) : false;
+
+    const requestErrorMsg = factChart.loaderStatus === STATUS.ERROR ? __('Request Failed') : undefined;
```
what do you think about moving `__('No data available')` to here instead of undefined?

amirfefer commented on webpack/assets/javascripts/react_app/redux/reducers/factCharts/factChart.fixtures.js:
```diff
@@ -0,0 +1,35 @@
+});
+
+export const afterModalOpen = Immutable({
+  modalToDisplay: { 1: true },
```
could you dry it a bit as well? some blocks are repeated here.

amirfefer commented on webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js:
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import FactChart from './';
+import Store from '../../redux';
+
```
could you use snapshot testing here?

amirfefer reviewed (changes requested):
Thanks @lizagilman ! 
almost there !
I left some inline comments.

lizagilman commented:
[test foreman]

lizagilman commented:
updated

amirfefer reviewed (changes requested):
LGTM! 
one last thing - could you please rename the commit title? 
it's not describe this PR.
something like : "Use pf-react donut chart in facts page" or so.
btw tests are red, try rebase?

lizagilman commented:
@amirfefer thanks, renamed
this branch is up to date, I think this is not related.

amirfefer commented:
@lizagilman the commit message still old

amirfefer reviewed (approved):
Thanks @lizagilman 
LGTM

ohadlevy commented on webpack/assets/javascripts/react_app/components/componentRegistry.js:
```diff
@@ -8,6 +8,8 @@ import ToastsList from './toastNotifications/';
 import StorageContainer from './hosts/storage/vmware/';
 import BookmarkContainer from './bookmarks';
 import PasswordStrength from './user/passwordStrength/';
+// eslint-disable-next-line
```
is this required?

lizagilman replied:
yes, otherwise there's an error - Using exported name 'FactChart' as identifier for default export
but I do want to use that name

ohadlevy replied:
i think this happens because you export twice in factCharts, i believe you only need to export the default one.

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/factChart.stories.js:
```diff
@@ -0,0 +1,37 @@
+import { storiesOf } from '@storybook/react';
+import thunk from 'redux-thunk';
+import configureMockStore from 'redux-mock-store';
+// eslint-disable-next-line
```
also not sure if its required?

lizagilman replied:
same here

ohadlevy replied:
same here too :)

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,90 @@
+import { STATUS } from '../../constants';
+import * as FactChartActions from '../../redux/actions/factCharts/';
+
+
```
extra whitespaces.

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,90 @@
+          key={`message-${id}`}
+        />) : false;
+
+
```
extra whitespaces - can you make sure you run those changes with prettier ?

ohadlevy reviewed (changes requested):
Thanks @lizagilman I've left a few comments inline, beside those I'm missing a few things:

1. clicking on the chart does not redirect to search.
2. I'm missing some UI (e.g. header, close "X" btn etc) - see the current screenshot below:

![image](https://user-images.githubusercontent.com/20167/37252220-2835f482-2526-11e8-8481-f6e327581823.png)

lizagilman commented:
@ohadlevy updated

ohadlevy commented:
@lizagilman please make sure to update the snapshots too once they are changed.

ohadlevy commented on .storybook/preview-head.html:
```diff
@@ -11,3 +11,12 @@
+<!-- workaround to get Jed into storybook -->
+
+<script>
+  const sprintfStr = (str, elem) => { return str } ;
```
nitpick: this is a static html / js script tag which does not go though babel, maybe we should not use ES6 here as it would limit storybook to recent browsers.

lizagilman commented:
@ohadlevy just noticed
updated

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js:
```diff
@@ -0,0 +1,30 @@
+
+const state = {
+  chartData: [],
+  loaderStatus: '',
```
why empty string? shouldn't this be a const of sort?

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js:
```diff
@@ -0,0 +1,30 @@
+
+    wrapper.simulate('click');
+    expect(props.onClick).toBeCalled();
+  });
```
missing tests for searching?

lizagilman replied:
shouldn't this functionality be tested at `ChartService`?

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,104 @@
+      false
+    );
+
+    const tooltip = factChart.chartData ? {
```
imho you should use react tooltip, not jquery

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,104 @@
+    );
+
+    const tooltip = factChart.chartData ? {
+      title: __('Show distribution chart'),
```
missing the host counter (see the screenshot i attached originally)

ohadlevy reviewed (changes requested):
please also delete the old view files / js as part of this pr.

ohadlevy commented:
@amirfefer please review again - thanks.

lizagilman commented:
@amirfefer @ohadlevy updated
@amirfefer can you confirm it is safe to delete `app/assets/javascripts/charts.js `?

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/factCharts/index.js:
```diff
@@ -0,0 +1,45 @@
+    case FACT_CHART_DATA_REQUEST:
+      return state.set('loaderStatus', 'PENDING');
+    case FACT_CHART_DATA_SUCCESS:
+      action.payload.values.forEach((val) => { (hostsCount += val[1]); });
```
this is probably a good candidate to introduce https://github.com/reactjs/reselect - @sharvit @amirfefer what do you guys think?

ohadlevy replied:
@sharvit @amirfefer @waldenraines @tstrachota  ping?

ohadlevy replied:
@lizagilman can you please create a follow up issue on redmine so we don't miss this?

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,111 @@
+      onclick: handleChartClick,
+    };
+
+    const chart = <DonutChart {...chartProps} config='large' />;
```
if the field has a long name, it's not readable, I'm not sure if its the c3 impementation (e.g. we need to open an issue on patternfly-react or somewhere else) but this its how it looks:
![image](https://user-images.githubusercontent.com/20167/37874288-3b557bfe-3035-11e8-87af-0fd3d86540d8.png)

vs

![image](https://user-images.githubusercontent.com/20167/37874290-47d329bc-3035-11e8-8e6a-44cd1883256b.png)

lizagilman replied:
I and @amirfefer think we should open an issue on that

ohadlevy commented on webpack/assets/javascripts/react_app/components/factCharts/index.js:
```diff
@@ -0,0 +1,111 @@
+                </b>
+                <small>
+                  {// eslint-disable-next-line no-undef
+                  Jed.sprintf(
```
please add parenthesis around the counter (e.g. like in this screenshot) 
![image](https://user-images.githubusercontent.com/20167/37874300-7a183d4a-3035-11e8-95f7-457fff0595c3.png)

lizagilman replied:
added

ohadlevy commented:
@lizagilman regarding removing charts.js / flot - please leave it and instead please fix http://projects.theforeman.org/issues/23010 - thanks :)

ohadlevy commented:
can we limit the text using ellipsis or similar until its resolved?

On Mon, Mar 26, 2018 at 1:30 PM, lizagilman <notifications@github.com>
wrote:

> *@lizagilman* commented on this pull request.
> ------------------------------
>
> In webpack/assets/javascripts/react_app/components/factCharts/index.js
> <https://github.com/theforeman/foreman/pull/5019#discussion_r177044666>:
>
> > +    const {
> +      factChart,
> +      data: { id, title, search },
> +      modalToDisplay,
> +    } = this.props;
> +
> +    const handleChartClick =
> +      search && search.match(/=$/) ? null : navigateToSearch.bind(null, search);
> +
> +    const chartProps = {
> +      data: factChart.chartData ? factChart.chartData : null,
> +      key: `chart-${id}`,
> +      onclick: handleChartClick,
> +    };
> +
> +    const chart = <DonutChart {...chartProps} config='large' />;
>
> I and @amirfefer <https://github.com/amirfefer> think we should open an
> issue on that
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5019#discussion_r177044666>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx5al1LytLRq-fk2f6SvDqvmaKZjpks5tiMNXgaJpZM4QmV4B>
> .
>

amirfefer commented:
@ohadlevy  as far as i know, we can pass trim values, but the tooltip won't show with the full value.
**EDIT**: after a quick look we can pass [tooltip.name](http://c3js.org/reference.html#tooltip-format-name-props) with the full value, and trim the values.

lizagilman commented:
@ohadlevy @amirfefer updated,
we now have the shortened labels and tooltips with full text 

![legend_tooltip](https://user-images.githubusercontent.com/15361156/38030571-a5a9e634-32a1-11e8-99a6-2deaab21ecc6.png)

![screenshot from 2018-03-28 16-00-27](https://user-images.githubusercontent.com/15361156/38030569-a583f96a-32a1-11e8-9982-450fbe51864a.png)

lizagilman commented:
@ohadlevy @amirfefer ping?

ohadlevy reviewed (dismissed):
thanks, this looks good, I've rebased and pushed so tests can run again, once they pass, I'll merge.

ohadlevy commented:
@amirfefer @sharvit any idea on why tests might be failing?

ohadlevy commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -30,15 +30,48 @@ const getChartConfig = ({
+
+  if (dataExists) {
+    dataWithShortNames = data.map((val) => {
+      const item = val.slice().asMutable();
```
note this breaks the dashboard page.

amirfefer commented:
@ohadlevy ,
Note that  there is a js error in the dashboard page, which caught by `dashboard_integration_test` .
Looks like this [change](https://github.com/theforeman/foreman/pull/5019/files#diff-1301a72a99b657c8c32ace90d1d5c670R42) causes this issue.

lizagilman commented:
[test foreman]
[test katello]

lizagilman commented:
[test foreman]

lizagilman commented:
[test foreman]

lizagilman commented:
@amirfefer @ohadlevy tests are green

lizagilman commented:
@amirfefer any other comments?

amirfefer commented:
there are some js warnings when a modal opens:
![screenshot from 2018-05-03 15-57-01](https://user-images.githubusercontent.com/11807069/39584100-4699d8aa-4efa-11e8-9e73-ac58b696b3ff.png)

I fixed this issue [here](https://github.com/theforeman/foreman/pull/5524)

amirfefer commented:
The modal opens with a full black background: 
![with-backdrop](https://user-images.githubusercontent.com/11807069/39586042-cb682f38-4efe-11e8-85df-226a1f14a583.png)

but when we pass `{backdrop={false}` the background is visible:
![with-no-backdrop](https://user-images.githubusercontent.com/11807069/39586052-cf0cc5ae-4efe-11e8-8254-a16b4ea4fdda.png)

@Rohoover, should the background be still visible?

Rohoover reviewed (commented):
@amirfefer 

It should neither be black or clear. 

See the PF storybook modal for the correct transparency for modals:

https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?selectedKind=About%20Modal&selectedStory=AboutModal&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

Sorry I don't have the transparency percentage off the top of my head.

lizagilman commented:
rebased, 
no warnings, and modal looks normal - half transparent (no black background)

amirfefer reviewed (approved):
Thanks! LGFM :+1:

lizagilman commented:
@tstrachota can we merge? any more comments?

ohadlevy commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -30,15 +31,48 @@ const getChartConfig = ({
       onclick,
       ...(colorsSize > 0 ? { colors } : {}),
     },
+    // eslint-disable-next-line
```
why did you need to disable all linting here? you can probably just use 
```js
    // eslint-disable-next-line no-shadow
````

lizagilman replied:
updated

ohadlevy reviewed (commented):
This looks fairly good, I think I have one last request, when opening / closing a modal, we always make an API call, maybe we can make the api call only if we don't have the data?

lizagilman commented:
I think this issue has come up already, someone mentioned that if we cache the data here users won't be able to see the most recent data, and they would want to get the updated data each time they open the modal. One thing I can think of is maybe we can add a refresh button inside the modal.

ohadlevy commented:
Ok, let's keep as is, in either cases I don't expect someone to stay long
enough in the index page that it would matter that much.

On Wed, May 9, 2018, 12:09 PM lizagilman <notifications@github.com> wrote:

> I think this issue has come up already, someone mentioned that if we cache
> the data here users won't be able to see the most recent data, and they
> would want to get the updated data each time they open the modal. One thing
> I can think of is maybe we can add a refresh button inside the modal.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5019#issuecomment-387675154>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx2k870wg6EenlJ7FZfA5wocBnu-Rks5twrJjgaJpZM4QmV4B>
> .
>
